apiclient/src/qs_api/mod.rs

```diff
@@ -77,43 +77,31 @@ impl ApiClient {
         request_params: QsRequestParamsOut,
         token_or_signing_key: AuthenticationMethod<'_, T>,
     ) -> Result<QsProcessResponseIn, QsRequestError> {
-        let tbs = ClientToQsMessageTbsOut::new(request_params);
-        let message = match token_or_signing_key {
-            AuthenticationMethod::Token(token) => ClientToQsMessageOut::from_token(tbs, token),
-            AuthenticationMethod::SigningKey(signing_key) => tbs
-                .sign(signing_key)
-                .map_err(|_| QsRequestError::LibraryError)?,
-            AuthenticationMethod::None => ClientToQsMessageOut::without_signature(tbs),
+        let api_version = self.negotiated_versions().qs_api_version();
+
+        let message = sign_params(request_params, &token_or_signing_key, api_version)?;
+        let endpoint = self.build_url(Protocol::Http, ENDPOINT_QS);
+        let response = send_qs_message(&self.client, &endpoint, &message).await?;
+
+        // check if we need to negotiate a new API version
+        let Some(accepted_version) = api_version_negotiation(
+            &response,
+            api_version,
+            ClientToQsMessageTbs::SUPPORTED_API_VERSIONS,
+        )
+        .transpose()?
+        else {
+            return process_response(response).await;
         };
-        let message_bytes = message
-            .tls_serialize_detached()
-            .map_err(|_| QsRequestError::LibraryError)?;
-        match self
-            .client
-            .post(self.build_url(Protocol::Http, ENDPOINT_QS))
-            .body(message_bytes)
-            .send()
-            .await
-        {
-            Ok(res) => {
-                let status = res.status();
-                if status.is_success() {
-                    // Success!
-                    let ds_proc_res_bytes = res.bytes().await.map_err(QsRequestError::Reqwest)?;
-                    let ds_proc_res = QsVersionedProcessResponseIn::tls_deserialize_exact_bytes(
-                        &ds_proc_res_bytes,
-                    )
-                    .map_err(QsRequestError::Tls)?;
-                    migrate_qs_process_response(ds_proc_res)
-                } else {
-                    // Error
-                    let error = res.text().await.map_err(QsRequestError::Reqwest)?;
-                    Err(QsRequestError::RequestFailed { status, error })
-                }
-            }
-            // A network error occurred.
-            Err(err) => Err(QsRequestError::NetworkError(err.to_string())),
-        }
+
+        self.negotiated_versions()
+            .set_qs_api_version(accepted_version);
```

Comment:
Do we want to allow version downgrades? If not, we should check (here or in `set_qs_api_version`, whether the new version is higher than the old one.

Reply:
Just to iterate, by downgrade you mean a newer Client connects to an older Server, right?

I just realized that we always respond with the current server API, currently. This is wrong. If we accept the API version of the client on the server, we also have to answer with that API version.